### PR TITLE
Fix App_offline and start adding tests

### PR DESCRIPTION
**Pull request by jkotalik**
Will work on adding a few more test cases. For #950

### CODE REVIEW

**Comment by pakrym** [comment]
Make sure this change doesn't cause process exit for inprocess TestServer

File: test/IISIntegration.FunctionalTests/Inprocess/AppOfflineTests.cs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        [ConditionalFact]
+        public async Task AppOfflineDropped_CanRemoveAppOfflineAfterAdding()
+        {
+            var deploymentResult = await StartSite();
+
+            AddAppOffline(deploymentResult.DeploymentResult.ContentRoot);
+
+            var response = await deploymentResult.RetryingHttpClient.GetAsync("HelloWorld");
+
+            while (response.IsSuccessStatusCode)
```
Comment (pakrym) [inline test/IISIntegration.FunctionalTests/Inprocess/AppOfflineTests.cs]: Possible infinite loop./

File: test/IISIntegration.FunctionalTests/Inprocess/AppOfflineTests.cs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        {
+            var fs = File.Create(Path.Combine(appPath, "app_offline.htm"));
+            // Close the file to allow for 
+            fs.Close();
+        }
+
+        private void RemoveAppOffline(string appPath)
+        {
+
+            RetryHelper.RetryOperation(
```
Comment (pakrym) [inline test/IISIntegration.FunctionalTests/Inprocess/AppOfflineTests.cs]: Do we lock appofline?

File: test/IISIntegration.FunctionalTests/Inprocess/AppOfflineTests.cs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+            var response = await deploymentResult.RetryingHttpClient.GetAsync("HelloWorld");
+
+            var responseText = await response.Content.ReadAsStringAsync();
+            Assert.Equal("Hello World", responseText);
+            return deploymentResult;
+        }
+
+        private void AddAppOffline(string appPath)
+        {
+            var fs = File.Create(Path.Combine(appPath, "app_offline.htm"));
```
Comment (pakrym) [inline test/IISIntegration.FunctionalTests/Inprocess/AppOfflineTests.cs]: File.WriteAllText?

File: test/IISIntegration.FunctionalTests/Inprocess/AppOfflineTests.cs [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+            AddAppOffline(deploymentResult.DeploymentResult.ContentRoot);
+            var response = await deploymentResult.RetryingHttpClient.GetAsync("HelloWorld");
+
+            while (response.IsSuccessStatusCode)
+            {
+                // Keep retrying until app_offline is present.
+                response = await deploymentResult.RetryingHttpClient.GetAsync("HelloWorld");
+            }
+
+            Assert.Equal(HttpStatusCode.ServiceUnavailable, response.StatusCode);
```
Comment (pakrym) [inline test/IISIntegration.FunctionalTests/Inprocess/AppOfflineTests.cs]: Verify that  app_offline is served.

File: test/IISIntegration.FunctionalTests/Inprocess/AppOfflineTests.cs [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+            RemoveAppOffline(deploymentResult.DeploymentResult.ContentRoot);
+        }
+
+        private async Task<IISDeploymentResult> StartSite()
+        {
+            var deploymentParameters = Helpers.GetBaseDeploymentParameters();
+
+            var deploymentResult = await DeployAsync(deploymentParameters);
+
+            var response = await deploymentResult.RetryingHttpClient.GetAsync("HelloWorld");
```
Comment (pakrym) [inline test/IISIntegration.FunctionalTests/Inprocess/AppOfflineTests.cs]: Too many hello worlds around. Let's add VerifyStarted helper and use it everywhere.

File: test/IISIntegration.FunctionalTests/Inprocess/AppOfflineTests.cs (hunk 2 above)
Comment (jkotalik) [reply test/IISIntegration.FunctionalTests/Inprocess/AppOfflineTests.cs]: I think we do to read the contents of the file.

Comment (pakrym) [reply test/IISIntegration.FunctionalTests/Inprocess/AppOfflineTests.cs]: Right, but we should set FILE_SHARE_READ | WRITE | DELETE while doing so.

**Comment by jkotalik** [comment]
🆙 📅 . I would also say app_offline is not really a testable scenario for the hostable webcore skew as it requires dropping app_offline in a bin local folder.

**Comment by pakrym** [comment]
I'm not worried about testing app_ofline on HWC, I'm worried that calling HWCShutdown which we do in test server would kill the process.

**Comment by jkotalik** [comment]
This wouldn't hit shutdown for HWC because exit is only called from recycle, not shutdown.

File: test/IISIntegration.FunctionalTests/Inprocess/AppOfflineTests.cs [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        }
+
+        private async Task AssertAppOffline(IISDeploymentResult deploymentResult, string expectedResponse = "")
+        {
+            var response = await deploymentResult.RetryingHttpClient.GetAsync("HelloWorld");
+
+            for (var i = 0; response.IsSuccessStatusCode && i < 5; i++)
+            {
+                // Keep retrying until app_offline is present.
+                response = await deploymentResult.RetryingHttpClient.GetAsync("HelloWorld");
```
Comment (pakrym) [inline test/IISIntegration.FunctionalTests/Inprocess/AppOfflineTests.cs]: The client is already retrying, do you need to retry even more?
How long does it take?

File: test/IISIntegration.FunctionalTests/Inprocess/AppOfflineTests.cs [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+                var response = await deploymentResult.RetryingHttpClient.GetAsync("HelloWorld");
+            }
+            catch (HttpRequestException)
+            {
+                // dropping app_offline will kill the process
+            }
+
+            var hostShutdownToken = deploymentResult.DeploymentResult.HostShutdownToken;
+
+            hostShutdownToken.WaitHandle.WaitOne(millisecondsTimeout: 1000);
```
Comment (pakrym) [inline test/IISIntegration.FunctionalTests/Inprocess/AppOfflineTests.cs]: Assert.True

File: test/IISIntegration.FunctionalTests/Inprocess/AppOfflineTests.cs [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+{
+    [SkipIfIISExpressSchemaMissingInProcess]
+    public class AppOfflineTests : IISFunctionalTestBase
+    {
+        [ConditionalFact]
+        public async Task AppOfflineDroppedWhileSiteIsDown_SiteReturns503()
+        {
+            var deploymentParameters = Helpers.GetBaseDeploymentParameters();
+
+            var deploymentResult = await DeployAsync(deploymentParameters);
```
Comment (pakrym) [inline test/IISIntegration.FunctionalTests/Inprocess/AppOfflineTests.cs]: I wish there was a way to assert that site is no started yet.

**Comment by pakrym** [comment]
Add a test where you delete app_offline and the start the app.

**Comment by pakrym** [comment]
What about outofproc?

File: test/IISIntegration.FunctionalTests/Inprocess/AppOfflineTests.cs (hunk 6 above)
Comment (jkotalik) [reply test/IISIntegration.FunctionalTests/Inprocess/AppOfflineTests.cs]: yeah no need for retrying.

File: test/IISIntegration.FunctionalTests/Inprocess/AppOfflineTests.cs (hunk 8 above)
Comment (jkotalik) [reply test/IISIntegration.FunctionalTests/Inprocess/AppOfflineTests.cs]: Yeah it would be a bit tricky.

**Comment by jkotalik** [comment]
I'll file an issue for out of proc once this is merged.

File: test/IISIntegration.FunctionalTests/Inprocess/AppOfflineTests.cs [hunk 9]
```diff
@@ -67,6 +67,18 @@ public class AppOfflineTests : IISFunctionalTestBase
             RemoveAppOffline(deploymentResult.DeploymentResult.ContentRoot);
         }
 
+        [ConditionalFact]
+        public async Task AddRemoveAppOffline_AppStarts()
+        {
+            var deploymentResult = await AssertStarts();
+
+            AddAppOffline(deploymentResult.DeploymentResult.ContentRoot);
+
```
Comment (pakrym) [inline test/IISIntegration.FunctionalTests/Inprocess/AppOfflineTests.cs]: Make sure app starts here and appofline is served

**Comment by pakrym** [comment]
> This wouldn't hit shutdown for HWC because exit is only called from recycle, not shutdown.

But Recycle is called from ExecuteApplication after main exits.

**Pull request merged**